ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -348,6 +377,296 @@ protected Map<DataNodeId, List<RemoteReplicaInfo>> selectReplicas(Map<DataNodeId
     return replicas;
   }
 
+  /**
+   * Generates group ids for active iteration for replicas, all replicas are mapped to one group id only
+   * Generates group ids for all datanodes for standby no progress groups
+   * All group ids are unique
+   * All replicas are mapped to particular group id and each datanode has one stand by remote replica group id
+   * so we make less number of parallel calls to remote hosts
+   * @param groupIdToRemoteReplicaMap {@link Map} this map will be populated with group ids to replicas map
+   * @param standByNoProgressGroupId {@link Map}  this map will be populated with datanode to standby group id
+   */
+  void generateGroupIdsForReplicas(Map<Integer, List<RemoteReplicaInfo>> groupIdToRemoteReplicaMap,
```

Comment:
There are too many maps mapping an int to some object here and takes too much time and effort to mentally keep track of these structures.

What is the goal of mapping an ID to a list of replicas ? 
Is it to organize the replicas of a node into active and standby lists ?
Can you define an abstraction like and possibly pass this around ?
```
class PeerNode {
  List activeReplicas;
  List standByReplicas;
  DataNode node;
  <some more variables>
  getters()
  setters()
}
```

Try to define a function with a return type that tells us instantly what the fn is returning. This fn here doesn't return anything and the return values are embedded in the arglist. Can you separate out the return value from this arglist ?

Try to avoid using mutable-booleans here and define some abstraction to pass around. That abstraction should have all the info easily accessible.

Rename standByNoProgress to just `standby` because that is enough to explain what the replica does - it stands by without any progress on data replication.

Lets not rush to merge this.